scripts/new_python_based_image.py

```diff
@@ -0,0 +1,555 @@
+
+    Args:
+        pipfile_path (str): The path to the Pipfile.
+        target_version (str): The target Python version to use with `pipenv lock`.
+    """
+    LOGGER.info(f"Running pipenv lock for '{pipfile_path}'")
+    env = os.environ.copy()
+    env["PIPENV_PIPFILE"] = os.path.basename(pipfile_path)
+
+    try:
+        result = subprocess.run(
+            ["pipenv", "lock", "--python", target_version],
+            cwd=os.path.dirname(pipfile_path),
+            check=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=env
+        )
+        LOGGER.debug(result.stdout.decode())
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(f"Error running pipenv lock for '{pipfile_path}'")
+        LOGGER.debug(e.stderr.decode())
+
+
+def manual_checks():
+    """
+    Provides a list of manual checks to perform after the script execution.
+
+    Returns:
+        list: A list of manual checks to review.
+    """
+    return [
+        "Check the issues thrown during the script execution, if any, and fix them manually.",
+        "Check if the Python version replacements have been performed correctly on the new files.",
+        "Review and make the appropriate changes in the Makefile and CI-related files.",
+        "Push the changes to a new branch on your fork to build the new images with GitHub workflows.",
+        "Test the new images."
+    ]
+
+
+def main():
+    """
+    Main function to execute the script.
+    """
+    args = extract_input_args()
+    LOGGER.info(args)
+
+    configure_logger(args.log_level)
+
+    with_logged_execution(lambda: check_requirements(args),
+                          "Checking requirements")
+
+    paths = with_logged_execution(lambda: find_matching_paths(args.context_dir, args.source, args.match),
+                                  f"Finding matching paths with '{args.match}' and Python {args.source}")
+
+    paths_dict = replace_python_version_on_paths(paths,
+                                                 args.source,
+                                                 args.target)
+
+    if len(paths_dict) == 0:
+        LOGGER.info("No paths found to copy.")
+        return
+
+    LOGGER.info(
+        f"New folders based on the input args:\n{dict_to_str(paths_dict, enumerate_lines=True)}")
+
+    success_paths, failed_paths = with_logged_execution(lambda: copy_paths(paths_dict),
+                                                        f"Trying to copy {len(paths_dict)} folders")
+
+    LOGGER.info(
+        f"{len(success_paths)} folders have been copied successfully whereas {len(failed_paths)} failed.")
+
+    if len(success_paths) > 0:
+        with_logged_execution(lambda: process_paths(success_paths, args.source, args.target),
+                              "Processing copied folders")
+
+    if len(failed_paths) > 0:
+        LOGGER.warning(
+            f"Failed to copy the following paths:\n{list_to_str(failed_paths)}")
```

Comment:
Again, is it okay to exit successfully once there are some failed paths?

Reply:
The idea is to continue the execution with the script for the other folders if this condition is met for a particular folder. Ultimately, you can see that this path has failed and act accordingly.

Reply:
Yes, I don't mean to stop the execution, but to exit with non-zero at the end... I guess that we will be running this manually, so it really doesn't matter that much, but in case of having this in some automation, then some non-zero exit code would be nice to realize that some of the paths failed to copy... otherwise one would have to grep for the warnings.